go/ics23.go

```diff
@@ -44,11 +50,84 @@ func VerifyMembership(spec *ProofSpec, root CommitmentRoot, proof *CommitmentPro
 // left and right proofs are neighbors (or left/right most if one is nil)
 // provided key is between the keys of the two proofs
 func VerifyNonMembership(spec *ProofSpec, root CommitmentRoot, proof *CommitmentProof, key []byte) bool {
-	// TODO: handle batch
-	np, ok := proof.Proof.(*CommitmentProof_Nonexist)
-	if !ok {
+	// decompress it before running code (no-op if not compressed)
+	proof = Decompress(proof)
+	np := getNonExistProofForKey(proof, key)
+	if np == nil {
 		return false
 	}
-	err := np.Nonexist.Verify(spec, root, key)
+	err := np.Verify(spec, root, key)
 	return err == nil
 }
+
+// BatchVerifyMembership will ensure all items are also proven by the CommitmentProof (which should be a BatchProof,
+// unless there is one item, when a ExistenceProof may work)
+func BatchVerifyMembership(spec *ProofSpec, root CommitmentRoot, proof *CommitmentProof, items map[string][]byte) bool {
+	// decompress it before running code (no-op if not compressed) - once for batch
+	proof = Decompress(proof)
+	for k, v := range items {
+		valid := VerifyMembership(spec, root, proof, []byte(k), v)
```

Comment:
Should there be another function or a flag to indicate contiguous keys for which some proof data could be shared?

Reply:
Well the proof data is shared automatically when you compress it. For any batch proof, whether it is range or not. Basically, if an inner node is present N times, it will be shrinked down to around (1 + N/16).

What it doesn't capture explicitly is the proof that all these N keys are continuous. There is the code to check `IsLeftNeighbor(spec, left, right)` which can be fed these items in batch and reconstruct they are a range.